### PR TITLE
[clang][ExprConst] allow single element access of vector object to be constant expression

### PR DESCRIPTION
**Pull request by vikramRH**
This is a slightly updated version of https://github.com/llvm/llvm-project/pull/72607 (I could not find a way to continue with the same PR, hopefully this is not very confusing). @sethp , I have tried addressing your views here although I have not given much thought on extending pathEntry for vector elements. 

originally authored by @yuanfang-chen (perhaps we could close the old PR?)

### CODE REVIEW

**Comment by vikramRH** [comment]
* **#101126** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/101126?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @vikramRH and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: clang/test/AST/Interp/vectors.cpp [hunk 1]
```diff
@@ -3,40 +3,40 @@
...
+static_assert(A[3] == 4, "");
 
 
 /// FIXME: It would be nice if the note said 'vector' instead of 'array'.
-static_assert(A[12] == 4, ""); // ref-error {{not an integral constant expression}} \
-                               // expected-error {{not an integral constant expression}} \
-                               // expected-note {{cannot refer to element 12 of array of 4 elements in a constant expression}}
+static_assert(A[12] == 4, ""); // both-error {{not an integral constant expression}} \
+                               // expected-note {{cannot refer to element 12 of array of 4 elements in a constant expression}} \
+                               // ref-note {{read of dereferenced one-past-the-end pointer is not allowed in a constant expression}}
```
Comment (tbaederr) [inline clang/test/AST/Interp/vectors.cpp:15]: Can you explain why this note is being generated? I don't see how `A[12]` is a one-past-the-end pointer.

File: clang/docs/ReleaseNotes.rst [hunk 2]
```diff
@@ -64,6 +64,9 @@ sections with improvements to Clang's support for those languages.
 
 C++ Language Changes
 --------------------
+- Allow single element access of vector object to be constant expression.
```
Comment (arsenm) [inline clang/docs/ReleaseNotes.rst:67]: This is potentially confusing to refer to "vector object". Be more specific this is an ext_vector_type vector?

File: clang/test/SemaCXX/constexpr-vectors-access-elements.cpp [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+
+using FourIntsExtVec __attribute__((ext_vector_type(4))) = int;
+
+constexpr FourIntsExtVec b = {1,2,3,4};
+static_assert(b[0] == 1 && b[1] == 2 && b[2] == 3 && b[3] == 4);
+static_assert(b.s0 == 1 && b.s1 == 2 && b.s2 == 3 && b.s3 == 4);
+static_assert(b.x == 1 && b.y == 2 && b.z == 3 && b.w == 4);
+static_assert(b.r == 1 && b.g == 2 && b.b == 3 && b.a == 4);
+static_assert(b[5]); // expected-error {{not an integral constant expression}} expected-note {{read of dereferenced one-past-the-end pointer}}
+
```
Comment (arsenm) [inline clang/test/SemaCXX/constexpr-vectors-access-elements.cpp:23]: Try access through an array and struct containing the vector

File: clang/lib/AST/ExprConstant.cpp [hunk 4]
```diff
@@ -442,6 +446,16 @@ namespace {
       MostDerivedArraySize = 2;
       MostDerivedPathLength = Entries.size();
     }
+
+    void addVectorElementUnchecked(QualType EltTy, uint64_t Size,
+                                   uint64_t Idx) {
+      Entries.push_back(PathEntry::ArrayIndex(Idx));
```
Comment (shafik) [inline clang/lib/AST/ExprConstant.cpp:452]: @sethp had a comment in the previous PR that is not address here: https://github.com/llvm/llvm-project/pull/72607#issuecomment-2175994595

**Review by shafik — COMMENTED** [review]
Thank you for working to get this PR finished up. I see at least two comments from the previous PR not addressed here. Please, let's try and get those addressed.

@AaronBallman wanted a test to make sure we reject `&V[0]` which I don't see: https://github.com/llvm/llvm-project/pull/72607#discussion_r1595550346

File: clang/lib/AST/ExprConstant.cpp (hunk 4 above)
Comment (vikramRH) [reply clang/lib/AST/ExprConstant.cpp:452]: Yes, I thought about having a new accessor of the sort "vectorIndex" but all it seems to achieve is just adding new API that returns does the exact same thing as array (other than perhaps adding a new meaning to PathEntry value). I will update it if you feel this makes sense.

File: clang/test/AST/Interp/vectors.cpp (hunk 1 above)
Comment (vikramRH) [reply clang/test/AST/Interp/vectors.cpp:15]: I just kept the original version of the PR, but the message "cannot refer to element 12 of array of 4 elements" seems correct here. I shall update this

Comment (vikramRH) [reply clang/test/AST/Interp/vectors.cpp]: done, let me know if the updated changes are okay

File: clang/docs/ReleaseNotes.rst (hunk 2 above)
Comment (vikramRH) [reply clang/docs/ReleaseNotes.rst]: done

File: clang/test/AST/Interp/vectors.cpp [hunk 5]
```diff
@@ -3,40 +3,39 @@
...
-static_assert(A[1] == 2, ""); // ref-error {{not an integral constant expression}}
-static_assert(A[2] == 3, ""); // ref-error {{not an integral constant expression}}
-static_assert(A[3] == 4, ""); // ref-error {{not an integral constant expression}}
+static_assert(A[0] == 1, "");
+static_assert(A[1] == 2, "");
+static_assert(A[2] == 3, "");
+static_assert(A[3] == 4, "");
 
 
 /// FIXME: It would be nice if the note said 'vector' instead of 'array'.
```
Comment (tbaederr) [inline clang/test/AST/Interp/vectors.cpp:12]: Resolving this fixme comment would be a nice follow-up.

**Comment by tbaederr** [comment]
If nobody else still has concerns, this LGTM from my end.

**Review by AaronBallman — APPROVED** [review]
LGTM!

**Comment by tbaederr** [comment]
@vikramRH Do you need someone else to merge this for you?

**Comment by vikramRH** [comment]
### Merge activity

* **Aug 7, 6:38 AM EDT**: @vikramRH started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/llvm/llvm-project/101126).

**Comment by vikramRH** [comment]
> @vikramRH Do you need someone else to merge this for you?

sorry for the delay, merged.

**Comment by ahatanak** [comment]
clang crashes compiling the following code.

```
typedef float float4 __attribute__((ext_vector_type(4)));
constexpr float4 v{1,2,3,4};
constexpr const float4 *p = &v;
constexpr float f = p->x;
```

I created a PR that fixes the crash: https://github.com/llvm/llvm-project/pull/136771

**Comment by ziqingluo-90** [comment]
```
typedef __attribute__((__ext_vector_type__(2))) float float2;

struct vec2 {
    float2 _v;
public:
    constexpr vec2(float x, float y) {
       _v.x = x;
       _v.y = y;
    }
};

constexpr struct vec2 f() {
  return vec2(1.0, 1.0);
}

int main() {
    vec2 S = f();
}
```
Clang crashes: https://godbolt.org/z/sx74s1bqd

**Comment by AaronBallman** [comment]
> ```
> typedef __attribute__((__ext_vector_type__(2))) float float2;
> 
> struct vec2 {
>     float2 _v;
> public:
>     constexpr vec2(float x, float y) {
>        _v.x = x;
>        _v.y = y;
>     }
> };
> 
> constexpr struct vec2 f() {
>   return vec2(1.0, 1.0);
> }
> 
> int main() {
>     vec2 S = f();
> }
> ```
> 
> Clang crashes: https://godbolt.org/z/sx74s1bqd

Can confirm the crash is happening. The issue is related to the constructor being `constexpr` and not initializing the member variable. If you use a member initializer list, the assert is silenced: https://godbolt.org/z/vEbeb4jbz

Can you file an issue so we don't lose track of the bug? Also @vikramRH any chance you can look into this?

**Comment by ziqingluo-90** [comment]
> Can you file an issue so we don't lose track of the bug?
Looks like we have an issue already #140348

@AaronBallman do you think my original example is well-defined according to the standard?  I saw there is a warning
`constexpr constructor that does not initialize all members is a C++20 extension [-Wc++20-extensions]`.

**Comment by tbaederr** [comment]
It's well defined, even if only after c++20. It should work fine, but the root of the problem is pre-existing (which is why it also fails with complex values).

**Comment by vikramRH** [comment]
@AaronBallman @tbaederr I can look into this, shall i assign https://github.com/llvm/llvm-project/issues/84315 to myself ?

**Comment by AaronBallman** [comment]
> @AaronBallman @tbaederr I can look into this, shall i assign #84315 to myself ?

Yup! FWIW, we don't require assignments in GitHub issue, you're always welcome to post a PR even if an issue isn't assigned to you.

**Comment by shafik** [comment]
This is also linked to the following crash: https://github.com/llvm/llvm-project/issues/220256

**Pull request merged**